future-apps/staking/test/staking.js

```diff
@@ -0,0 +1,480 @@
+      const kernelBase = await getContract('Kernel').new()
+      const aclBase = await getContract('ACL').new()
+      daoFact = await getContract('DAOFactory').new(kernelBase.address, aclBase.address, regFact.address)
+    })
+
+    beforeEach(async () => {
+      // scripts
+      executionTarget = await getContract('ExecutionTarget').new()
+      const stakeAction = { to: executionTarget.address, calldata: executionTarget.contract.executeStake.getData() }
+      const stakeScript = encodeCallScript([stakeAction])
+      const unstakeAction = { to: executionTarget.address, calldata: executionTarget.contract.executeUnstake.getData() }
+      const unstakeScript = encodeCallScript([unstakeAction])
+      const lockAction = { to: executionTarget.address, calldata: executionTarget.contract.executeLock.getData() }
+      const lockScript = encodeCallScript([lockAction])
+
+      // token
+      token = await getContract('StandardTokenMock').new(owner, balance)
+
+      // DAO
+      const r = await daoFact.newDAO(owner)
+      const dao = getContract('Kernel').at(r.logs.filter(l => l.event == 'DeployDAO')[0].args.dao)
+      const acl = getContract('ACL').at(await dao.acl())
+
+      await acl.createPermission(owner, dao.address, await dao.APP_MANAGER_ROLE(), owner, { from: owner })
+
+      // App
+      const receipt = await dao.newAppInstance('0x1234', (await getContract('Staking').new()).address, { from: owner })
+      app = getContract('Staking').at(receipt.logs.filter(l => l.event == 'NewAppProxy')[0].args.proxy)
+      await app.initialize(token.address, stakeScript, unstakeScript, lockScript)
+
+      // Permissions
+      await acl.createPermission(owner, app.address, await app.STAKE_ROLE(), owner, { from: owner })
+      await acl.createPermission(owner, app.address, await app.UNSTAKE_ROLE(), owner, { from: owner })
+      await acl.createPermission(owner, app.address, await app.LOCK_ROLE(), owner, { from: owner })
+    })
+
+    it('stakes and runs script', async () => {
+      const amount = 100
+      const initialOwnerBalance = parseInt((await token.balanceOf(owner)).valueOf(), 10)
+      const initialStakingBalance = parseInt((await token.balanceOf(app.address)).valueOf(), 10)
+
+      // allow Staking app to move owner tokens
+      await token.approve(app.address, amount)
+      // stake tokens
+      await app.stake(amount, '')
+
+      const finalOwnerBalance = parseInt((await token.balanceOf(owner)).valueOf(), 10)
+      const finalStakingBalance = parseInt((await token.balanceOf(app.address)).valueOf(), 10)
+      assert.equal(finalOwnerBalance, initialOwnerBalance - amount, "owner balance should match")
+      assert.equal(finalStakingBalance, initialStakingBalance + amount, "Staking app balance should match")
+      assert.equal((await app.totalStakedFor(owner)).valueOf(), amount, "staked value should match")
+      assert.equal(await executionTarget.stakeCounter(), 1, 'should have received execution call')
+    })
+
```

Comment:
Add test for stakeAndLock triggering both scripts

Reply:
Done.